chanfitness/chanevent.go

```diff
@@ -0,0 +1,179 @@
+}
+
+// chanEventLog stores all events that have occurred over a channel's lifetime.
+type chanEventLog struct {
+	// id is the uint64 of the short channel ID.
+	id uint64
+
+	// peer is the compressed public key of the peer being monitored.
+	peer [33]byte
+
+	// events is a log of timestamped events observed for the channel.
+	events []*channelEvent
+
+	// now is expected to return the current time. It is supplied as an
+	// external function to enable deterministic unit tests.
+	now func() time.Time
+
+	// closed indicates whether a channel is closed, and is used to determine
+	// whether we should continue to track peer events for the channel without
+	// having to iterate through all events. When channel scored are persisted,
+	// closed channels can be removed from memory rather than having a marker.
+	closed bool
+}
+
+// add appends an event with the given type and current time to the event log.
+func (e *chanEventLog) add(eventType eventType) {
+	e.events = append(e.events, &channelEvent{
+		timestamp: e.now(),
+		eventType: eventType,
+	})
+
+	if eventType == channelClosedEvent {
+		e.closed = true
+	}
+
+	log.Debugf("Channel %v recording event: %v", e.id, eventType)
+}
+
+// getLifespan returns the total observed lifetime of a channel's event log.
+// If a channel is open, the lifespan is calculated from open time until the
+// present. If the channel is closed, this lifespan terminates when the channel
+// was closed. In absence of a channel opened event, the earliest entry in the
+// event log is considered to be the opening time.
+func (e *chanEventLog) getLifespan() time.Duration {
+	// If there are no events, we cannot calculate lifetime.
+	if len(e.events) == 0 {
+		return 0
+	}
+
+	// Close time starts as the present time, but will be set back to the actual
+	// closing time if there is a close event present.
+	closeTime := e.now()
+
+	var openTime time.Time
+
+	for _, e := range e.events {
+		// Set open time to the earliest timestamp in the event log. The earliest
+		// event is expected to be a channel opened event, but if it is not, it
+		// will be overridden by a subsequent open event, should one be present,
+		// because the events log is ordered by time for this loop.
+		if openTime.IsZero() || e.timestamp.Before(openTime) {
+			openTime = e.timestamp
+		}
+
+		switch e.eventType {
+		case channelOpenedEvent:
+			openTime = e.timestamp
+
+		case channelClosedEvent:
+			closeTime = e.timestamp
+		}
+
+	}
+
+	// If a close event was recorded before an open event, just return zero to
+	// prevent returning a negative lifespan.
+	if closeTime.Before(openTime) {
+		return 0
+	}
+
+	return closeTime.Sub(openTime)
+}
+
+// getUptime returns total uptime of a channel. Online periods are calculated
+// as the time between the node reporting as online, then subsequently
+// being reported offline. Any out of state events, such as two adjacent online
+// events are ignored, and the earlier timestamp is used for uptime calculation.
```

Comment:
They shouldn't be added to the log at all, so this really only covers the case where the peer subscription has a bug and sends two online events. I can add a warning log? Not sure what the go-to is for when lnd gets something unexpected like this?

Reply:
If it can only be caused by a bug, normally we'd return an error to the caller. To handle it at the origin. But in this case we are building a reporting system which doesn't directly influence the rest of the system. And also there is no return path available yet. I think logging it as error (in case this really should never happen) is the best option.